application_base.cpp

```diff
@@ -191,19 +191,19 @@ void application::startup() {
 
    //after startup, shut down the signal handling thread and catch the signals back on main io_service
    clean_up_signal_thread();
-   setup_signal_handling_on_ios(get_io_service(), false);
+   setup_signal_handling_on_ios(io_serv, false);
 
 #ifdef SIGHUP
-   std::shared_ptr<boost::asio::signal_set> sighup_set(new boost::asio::signal_set(get_io_service(), SIGHUP));
+   std::shared_ptr<boost::asio::signal_set> sighup_set(new boost::asio::signal_set(io_serv, SIGHUP));
    start_sighup_handler( sighup_set );
 #endif
 }
 
-void application::start_sighup_handler( std::shared_ptr<boost::asio::signal_set> sighup_set ) {
+void application_base::start_sighup_handler( std::shared_ptr<boost::asio::signal_set> sighup_set ) {
 #ifdef SIGHUP
    sighup_set->async_wait([sighup_set, this](const boost::system::error_code& err, int /*num*/) {
       if( err ) return;
-      app().post(priority::medium, [sighup_set, this]() {
+      post_cb(priority::medium, [sighup_set, this]() {
```

Comment:
Why was this changed? `app().post` still works here. Although, I would argue it should be changed to `app().executor().post`

Reply:
agreed, ot should be `app().post()`.

It cannot be `app().executor().post` as the appbase user may change the executor() and we don't want the executor type hardcoded in appbase.

Reply:
I'm starting to think we should divorce executor from application. The executor could be provided to application where needed and users just interact with an executor outside of app()

Reply:
`application_base` uses the executor. If we have a separation, it will require a virtual interface for the executor which would be slower.

Reply:
It could be passed into `exec()`.

Reply:
The issue with this is that the user could pass an executor to exec(), and a different one to the other `application` APIs. I think the current implementation is safer, and I don't see the benefit of a change

Reply:
But yes, maybe we should make `application` a template as you suggested, instead of passing the type in `executor_t`. Let me try that.

Reply:
@heifner I just updated `application` to be a template as you suggested.